[Entry Edit] Python Built-in Functions: `breakpoint()` (#1301)
Codecademy/docs: content/python/concepts/built-in-functions/terms/breakpoint/breakpoint.md

  - 'paths/data-science'
---

The **`breakpoint()`** function engages, configures, and changes the debugger program used in a script. 

When a breakpoint is reached, it triggers a pause in the execution of the program, allowing the programmer to examine and debug the current state of the code. This feature can be very helpful in trying to figure out why something isn't working correctly or getting an understanding of what is happening. With breakpoints, the programmer can step through the code line-by-line, and watch the output as they progress through the program.

## Syntax

```pseudo
breakpoint()
```

By default, the `breakpoint()` function imports the `pdb` debugger module.
`breakpoint()` was introduced in [Python 3.7](https://docs.python.org/3/whatsnew/3.7.html#whatsnew37-pep553) via [PEP 553](https://peps.python.org/pep-0553/). Earlier versions of Python required explicitly setting up a debugger using the following syntax:

```pseudo
import pdb; pdb.set_trace()
```

> So, why did they introduce the `breakpoint()` function? For one, it is much more concise and easier to type. Plus, it offers some additional features, such as the ability to choose different debugging tools.

Author: No, it offers the possibility to change the Debugger Module. For example change it to the web-pdb module.

  - 'paths/data-science'
---

The **`breakpoint()`** function engages, configures, and changes the debugger program used in a script. 

> When a breakpoint is reached, it triggers a pause in the execution of the program, allowing you to examine and debug the current state of your code. This feature can be very helpful if you are trying to figure out why something isn't working correctly or just want to get an understanding of what is happening. With breakpoints, you can step through the code line-by-line, and watch the output as you progress through the program.

Reviewer: ```suggestion
When a breakpoint is reached, it triggers a pause in the execution of the program, allowing the programmer to examine and debug the current state of the code. This feature can be very helpful in trying to figure out why something isn't working correctly or getting an understanding of what is happening. With breakpoints, the programmer can step through the code line-by-line, and watch the output as they progress through the program.
```

Reviewer: Note we try to avoid direct address to the reader, so no "you" or "we".

## Syntax

```pseudo
breakpoint()
```

By default, the `breakpoint()` function imports the `pdb` debugger module.

> `breakpoint()` was introduced in [Python 3.7](https://docs.python.org/3/whatsnew/3.7.html#whatsnew37-pep553) via [PEP 553](https://peps.python.org/pep-0553/). In earlier versions of Python, you had to explicitly set up a debugger using the following syntax:

Reviewer: ```suggestion
`breakpoint()` was introduced in [Python 3.7](https://docs.python.org/3/whatsnew/3.7.html#whatsnew37-pep553) via [PEP 553](https://peps.python.org/pep-0553/). Earlier versions of Python required explicitly setting up a debugger using the following syntax:
```

```pseudo
import pdb; pdb.set_trace()
```

So, why did they introduce the `breakpoint()` function? For one, it is much more concise and easier to type. Plus, it offers some additional features, such as the ability to choose different debugging tools.

### Common `pdb` Debugger Commands

> But by default, you'll be working with the `pdb` debugger. Here are some common commands you can use while debugging:

Reviewer: ```suggestion
The default is to work with `pdb` debugger. Here are some common commands used while debugging:
```

> - `h` (help): Print a list of available commands

Reviewer: Eliminate line breaks between items in the bulleted list.

Author: Done!

- `s` (step): Execute the current line, stop at the first possible occasion (either in a function that is called or on the next line in the current function).

-  `n` (next): Continue execution until the next line in the current function is reached, or it returns. The difference between next and step is that step stops inside a called function, while next executes called functions, only stopping at the next line in the current function.

- `c` (continue): Continue execution, only stop when a breakpoint is encountered.

> - `w` (where): Print a stack trace, with the most recent frame at the bottom. That means  you'll see the code that is currently executing and its callers.

Reviewer: ```suggestion
- `w` (where): Print a stack trace, with the most recent frame at the bottom, showing the code that is currently executing and its callers.
```

- `l` (list): Show the current location in the file and a few lines (by default 11 lines) of context.

- `q` (quit): Quit from the debugger. 

> For a full list of commands and more information, you can check out [the official documentation on Python's website](https://docs.python.org/3/library/pdb.html#debugger-commands).

Reviewer: ```suggestion
A full list of commands and more information are in [the official documentation on Python's website](https://docs.python.org/3/library/pdb.html#debugger-commands).
```

print("Welcome back to Codecademy!!")
```

> This is what happens when running this code in the terminal:

Reviewer 2: ```suggestion
This will cause the following in the terminal:
```

![Screenshot of terminal showing the breakpoint() formula](https://raw.githubusercontent.com/Codecademy/docs/main/media/python-built-in-functions-breakpoint-screenshot.png)

> The execution of the program was paused, showing a prompt in the terminal. At this point the code can be analyzed using the previously mentioned commands.

Reviewer 2: ```suggestion
The execution of the program was paused, showing a prompt in the terminal. At this point, the code can be analyzed using the previously mentioned commands.
```

print("Welcome back to Codecademy!!")
```

Let's take a look at what happens when you run this code in your terminal:

> ![Screenshot of terminal showing the breakpoint() formula](Screenshot-Terminal-Breakpoint.png)

Reviewer: Images should be in the Docs\media folder.

Author: @SSwiniarski I did so in my last commit. One thing I'm not sure about is the image link. Using ../../media/python-built-in-functions-breakpoint-screenshot.png didn't work, so I looked at other .md files that contain an image in the media folder and used the link they're using (but adapted it to my filename). Could you check if that works? In the Codecademy Docs style guide I couldn't find anything on images.

Reviewer: That should work.

> As you can see, the execution of our program was paused, and you get a prompt in the terminal. At this point, you can start analyzing your code using the previously mentioned commands.

Reviewer: ```suggestion
The execution of the program was paused, showing a prompt in the terminal. At this point the code can be analyzed using the previously mentioned commands.
```
